src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsCache.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.UnifiedSuggestions;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions
+{
+    /// <summary>
+    /// Caches suggested action sets between calls to <see cref="CodeActionsHandler"/> and
+    /// <see cref="CodeActionResolveHandler"/>.
+    /// </summary>
+    [Export(typeof(CodeActionsCache)), Shared]
+    internal class CodeActionsCache
+    {
+        /// <summary>
+        /// Ensures we aren't making concurrent modifications to the list of cached items.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        /// <summary>
+        /// Maximum number of cached items.
+        /// </summary>
+        private const int MaxCacheSize = 3;
+
+        /// <summary>
+        /// Current list of cached items.
+        /// </summary>
+        private readonly List<CodeActionsCacheItem> _cachedItems = new List<CodeActionsCacheItem>();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CodeActionsCache()
+        {
+        }
+
+        public async Task UpdateActionSetsAsync(
+            Document document,
+            LSP.Range range,
+            ImmutableArray<UnifiedSuggestedActionSet> cachedSuggestedActionSets,
+            CancellationToken cancellationToken)
+        {
+            using (await _semaphore.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                // If there's a value in the cache with the same document and range we're searching for,
+                // remove and replace it with our updated value.
+                var previousCachedItem = _cachedItems.Where(c => IsMatch(document, range, c));
+                if (!previousCachedItem.IsEmpty())
+                {
+                    _cachedItems.Remove(previousCachedItem.First());
+                }
```

Comment:
```suggestion
                var previousCachedItem = _cachedItems.FirstOrDefault(c => IsMatch(document, range, c));
                if (previousCachedItem != null)
                {
                    _cachedItems.Remove(previousCachedItem);
                }
```

Reply:
Thanks for your review!

Oops, I forgot to mention this in the last round of feedback, but I tried this syntax but it doesn't seem to work since `_cachedItems` is made up of struct items, so `FirstOrDefault` won't ever evaluate to null.
I guess we could make the struct a class instead? But that seems unnecessary as well, hmm...

Reply:
we also have FirstOrNull for struct cases.

Reply:
Oops, sorry, I completely missed that it was a struct!